File: ConsoleApp2/ConsoleApp2/Extensions/ServiceCollectionExtensions.vb
Imports Microsoft.Extensions.Configuration
Imports Microsoft.Extensions.DependencyInjection
Imports Microsoft.IdentityModel.Tokens
Imports Microsoft.OpenApi.Models
Imports System.Text
Imports System.Text.Json.Serialization
Imports System.Text.Json.Serialization.JsonConverter
Imports System.Text.Json.Serialization.ReferenceHandler
Imports System
Imports System.Collections.Generic
Imports infrastructure

Public Module ServiceCollectionExtensions
    <Runtime.CompilerServices.Extension>
    Public Sub AddAPIServices(services As IServiceCollection, config As IConfiguration)
        services.AddControllers().AddJsonOptions(
            Sub(opt) opt.JsonSerializerOptions.ReferenceHandler = ReferenceHandler.IgnoreCycles)

        ' Add services to the container.

        Dim tokenOptions = config.GetSection("TokenOptions").Get(Of TokenOptions)()

        services.AddAuthentication(
            Sub(opt)
                opt.DefaultAuthenticateScheme = "Bearer"
                opt.DefaultChallengeScheme = "Bearer"
            End Sub
        ).AddJwtBearer(
            Sub(opt)
                opt.TokenValidationParameters = New TokenValidationParameters() With
                {
                    .ValidateIssuer = True,
                    .ValidIssuer = tokenOptions.Issuer,
                    .ValidateAudience = True,
                    .ValidAudience = tokenOptions.Audience,
                    .ValidateLifetime = True,
                    .ValidateIssuerSigningKey = True,
                    .IssuerSigningKey = New SymmetricSecurityKey(Encoding.UTF8.GetBytes(tokenOptions.SecurityKey)),
                    .ClockSkew = TimeSpan.Zero
                }
            End Sub
        )

        ' Swagger/OpenAPI configuration
        services.AddEndpointsApiExplorer()
        services.AddSwaggerGen(
            Sub(c)
                c.AddSecurityDefinition("Bearer", New OpenApiSecurityScheme() With
                {
                    .Description = "JWT Authorization",
                    .Name = "Authorization",
                    .In = ParameterLocation.Header,
                    .Type = SecuritySchemeType.Http,
                    .Scheme = "Bearer",
                    .BearerFormat = "JWT"
                })

                c.AddSecurityRequirement(New OpenApiSecurityRequirement() From
                {
                    {
                        New OpenApiSecurityScheme With
                        {
                            .Reference = New OpenApiReference With
                            {
                                .Type = ReferenceType.SecurityScheme,
                                .Id = "Bearer"
                            }
                        },
                        New List(Of String)()
                    }
                })
            End Sub
        )
    End Sub
End Module

File: ConsoleApp2/ApiDeneme/Extensions/ServiceCollectionExtensions.cs
﻿
using infrastructure;
using Microsoft.IdentityModel.Tokens;
using Microsoft.OpenApi.Models;
using System.Text;
using System.Text.Json.Serialization;

namespace LM.WebAPI.Extensions
{
    public static class ServiceCollectionExtensions
    {
        public static void AddAPIServices(this IServiceCollection services, IConfiguration config)
        {
            services.AddControllers()
                .AddJsonOptions(opt => opt.JsonSerializerOptions.ReferenceHandler = ReferenceHandler.IgnoreCycles);

            // Add services to the container.

            //var tokenOptions = config.GetSection("TokenOptions").Get<TokenOptions>();

            //services.AddAuthentication(opt =>
            //{
            //    opt.DefaultAuthenticateScheme = "Bearer";
            //    opt.DefaultChallengeScheme = "Bearer";
            //}).AddJwtBearer(opt =>
            //{
            //    opt.TokenValidationParameters = new TokenValidationParameters()
            //    {
            //        ValidateIssuer = true,
            //        ValidIssuer = tokenOptions.Issuer,
            //        ValidateAudience = true,
            //        ValidAudience = tokenOptions.Audience,
            //        ValidateLifetime = true,
            //        ValidateIssuerSigningKey = true,
            //        IssuerSigningKey = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(tokenOptions.SecurityKey)),
            //        ClockSkew = TimeSpan.Zero
            //    };
            //});




            //builder.Services.AddControllers();
            // Learn more about configuring Swagger/OpenAPI at https://aka.ms/aspnetcore/swashbuckle
            services.AddEndpointsApiExplorer();
            services.AddSwaggerGen(c =>
            {
                c.AddSecurityDefinition("Bearer", new OpenApiSecurityScheme()
                {
                    Description = "JWT Auhorization",
                    Name = "Authorization",
                    In = ParameterLocation.Header,
                    Type = SecuritySchemeType.Http,
                    Scheme = "Bearer",
                    BearerFormat = "JWT"
                });

                c.AddSecurityRequirement(new OpenApiSecurityRequirement()
                {
                    {
                        new OpenApiSecurityScheme
                        {
                                Reference = new OpenApiReference
                                {
                                    Type = ReferenceType.SecurityScheme,
                                    Id="Bearer"
                                }
                        },

                        new List<string>()
                    }
                });
            });
        }
    }
}
